src/libraries/Microsoft.PowerFx.Core/Utils/MutationUtils.cs

```diff
@@ -0,0 +1,82 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+using System.Linq;
+using Microsoft.PowerFx.Core.App.ErrorContainers;
+using Microsoft.PowerFx.Core.Binding;
+using Microsoft.PowerFx.Core.Entities;
+using Microsoft.PowerFx.Core.Errors;
+using Microsoft.PowerFx.Core.Functions;
+using Microsoft.PowerFx.Core.Localization;
+using Microsoft.PowerFx.Core.Types;
+using Microsoft.PowerFx.Syntax;
+using Microsoft.PowerFx.Types;
+
+namespace Microsoft.PowerFx.Core.Utils
+{
+    internal class MutationUtils
+    {
+        public static void CheckForReadOnlyFields(DType dataSourceType, TexlNode[] args, DType[] argTypes, IErrorContainer errors)
+        {
+            if (dataSourceType.AssociatedDataSources.Any())
+            {
+                var tableDsInfo = dataSourceType.AssociatedDataSources.Single();
+
+                if (tableDsInfo is IExternalCdsDataSource cdsTableInfo)
+                {
+                    for (int i = 0; i < argTypes.Length; i++)
+                    {
+                        if (!cdsTableInfo.IsArgTypeValidForMutation(argTypes[i], out var invalidFieldNames))
+                        {
+                            errors.EnsureError(DocumentErrorSeverity.Severe, args[i], TexlStrings.ErrRecordContainsInvalidFields_Arg, string.Join(", ", invalidFieldNames));
+                        }
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Adds specific errors for mutation functions.
+        /// </summary>
+        public static void CheckSemantics(TexlBinding binding, TexlFunction function, TexlNode[] args, DType[] argTypes, IErrorContainer errors)
+        {
+            var targetArg = args[0];
+
+            // control!property are not valid targets for Collect, Remove, etc.
```

Comment:
>control!property [](http://example.com/codeflow?start=15&length=16)

`control.property` - we deprecated the `!` operator

Reply:
This was copied from PA.

Reply:
In Power Apps we still support the `!` operator for legacy purposes; in Power Fx (C#) we don't.